linkerd/proxy/http/src/insert.rs

```diff
@@ -1,20 +1,38 @@
+use futures::{Future, TryFuture};
 use linkerd_stack::{layer, NewService, Param, Proxy};
-use std::marker::PhantomData;
-use std::task::{Context, Poll};
+use std::{
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
 
 pub trait Lazy<V>: Clone {
     fn value(&self) -> V;
 }
 
-/// Wraps an HTTP `Service` so that the `T -typed value` is cloned into
-/// each request's extensions.
-#[derive(Clone, Debug)]
-pub struct Layer<L, V> {
+/// Wraps an HTTP `Service` so that a `P`-typed `Param` is cloned into each
+/// request's extensions.
+#[derive(Debug)]
+pub struct NewInsert<P, N> {
+    inner: N,
+    _marker: PhantomData<fn() -> P>,
+}
+
+/// Wraps an HTTP `Service` so that a `P`-typed `Param` is cloned into each
+/// response's extensions.
+#[derive(Debug)]
+pub struct NewResponseInsert<P, N> {
+    inner: N,
```

Comment:
nit: these should _probably_ have manual `Debug` implementations so that they don't require `P: Debug`...but, this probably doesn't actually matter